stdlib/public/SwiftShims/Visibility.h

```diff
@@ -101,41 +101,57 @@
 #if defined(__MACH__) || defined(__wasi__)
 
 # define SWIFT_EXPORT_ATTRIBUTE __attribute__((__visibility__("default")))
+# define SWIFT_IMPORT_ATTRIBUTE __attribute__((__visibility__("default")))
 
 #elif defined(__ELF__)
 
 // We make assumptions that the runtime and standard library can refer to each
 // other's symbols as DSO-local, which means we can't allow the dynamic linker
 // to relocate these symbols. We must give them protected visibility while
 // building the standard library and runtime.
-# if defined(swiftCore_EXPORTS)
-#  define SWIFT_EXPORT_ATTRIBUTE __attribute__((__visibility__("protected")))
-# else
-#  define SWIFT_EXPORT_ATTRIBUTE __attribute__((__visibility__("default")))
-# endif
+# define SWIFT_EXPORT_ATTRIBUTE __attribute__((__visibility__("protected")))
+# define SWIFT_IMPORT_ATTRIBUTE __attribute__((__visibility__("default")))
 
 // FIXME: this #else should be some sort of #elif Windows
 #else // !__MACH__ && !__ELF__
 
 # if defined(__CYGWIN__)
 #  define SWIFT_EXPORT_ATTRIBUTE
+#  define SWIFT_IMPORT_ATTRIBUTE
 # else
-
-#  if defined(swiftCore_EXPORTS)
-#   define SWIFT_EXPORT_ATTRIBUTE __declspec(dllexport)
-#  else
-#   define SWIFT_EXPORT_ATTRIBUTE __declspec(dllimport)
-#  endif
-
+#  define SWIFT_EXPORT_ATTRIBUTE __declspec(dllexport)
+#  define SWIFT_IMPORT_ATTRIBUTE __declspec(dllimport)
 # endif
 
 #endif
 
+// Make sure that SWIFT_LIBRARY_IS is defined for all the
+// libraries we care about.
+#ifndef SWIFT_LIBRARY_IS_swiftCore
+#define SWIFT_LIBRARY_IS_swiftCore 0
+#endif
+
+#define SWIFT_MACRO_CONCAT(A, B) A ## B
+#define SWIFT_MACRO_IF_0(IF_TRUE, IF_FALSE) IF_FALSE
+#define SWIFT_MACRO_IF_1(IF_TRUE, IF_FALSE) IF_TRUE
+#define SWIFT_MACRO_IF(COND, IF_TRUE, IF_FALSE) \
+  SWIFT_MACRO_CONCAT(SWIFT_MACRO_IF_, COND)(IF_TRUE, IF_FALSE)
```

Comment:
Wow, this really seems like a lot of gymnastics to avoid defining a couple of macros.  Is there a bigger motivation behind this rather than just trying to really avoid a second definition of a macro which expands to `__attribute__((__visibility__("default")))` and `__attribute__((__visibility__(("protected")))`?

Reply:
I want to avoid repeating a huge `#if` tree for every new runtime library we define.  Producing the target-specific export/import attributes is the biggest chunk of that, but since we also conditionally add `extern "C"`, we'd still have quite a bit of boilerplate repeated for each library.  I think a small amount of macro magic is warranted.